research/utils/log_operations.py

```diff
@@ -0,0 +1,101 @@
+import logging
+import struct, pdb, os
+import numpy as np
+import torch
+
+from PIL import Image
+from enum import Enum
+from typing import Callable, List, Dict
+from ignite.engine import Events, Engine
+from utils.helpers import tern
+from utils.image_utils import convert_cwh_to_whc
+
+
+def _to_stdout(
+    engine: Engine, fields: List[str], engine_attr: str, epoch_num=None, iter_num=None
+) -> None:
+    """Prints string formatted engine output fields to stdout"""
+    value_dict = getattr(engine.state, engine_attr)
+
+    log_str = "  ".join(
+        ["{}: {:.3f}".format(field, value_dict[field]) for field in fields]
+    )
+    log_str = "Epoch[{:d}], Iter[{:d}] | {}".format(
+        tern(epoch_num, engine.state.epoch),
+        tern(iter_num, engine.state.iteration),
+        log_str,
+    )
+    print(log_str)
+
+
+def _to_log(
+    engine: Engine, fields: List[str], engine_attr: str, epoch_num=None, iter_num=None,
+) -> None:
+    """Logs string formatted engine output fields to logfile"""
+    value_dict = getattr(engine.state, engine_attr)
+
+    log_str = "  ".join(
+        ["{}: {:.3f}".format(field, value_dict[field]) for field in fields]
+    )
+    log_str = "Epoch[{:d}], Iter[{:d}] | {}".format(
+        tern(epoch_num, engine.state.epoch),
+        tern(iter_num, engine.state.iteration),
+        log_str,
+    )
+    engine.logger.info(log_str)
+
+
+def _to_file(engine: Engine, fields: List[str], engine_attr: str) -> None:
```

Comment:
What is the main difference between logging to a log file and a separate binary file?

Reply:
The difference is a message in a log file (e.g. "Epoch[1] accuracy: ...", versus a binary file that stores the data (that you could load into Numpy)

Reply:
oh like some cvs? where it could be loaded into some table

Reply:
exactly. but storing as text (csv) is less efficient than binary